File: backend/app/schemas/playback_event.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


PlaybackEventType = Literal["play", "pause", "resume", "seek", "skip", "complete"]


class PlaybackEventSyncItem(BaseModel):
    client_event_id: str = Field(min_length=1, max_length=128)
    track_id: int
    event_type: PlaybackEventType
    position_seconds: float = Field(ge=0)
    duration_seconds: float | None = Field(default=None, ge=0)
    occurred_at: datetime
    client: str = Field(min_length=1, max_length=50)


class PlaybackEventBulkSyncRequest(BaseModel):
    events: list[PlaybackEventSyncItem] = Field(min_length=1, max_length=100)


class PlaybackEventAccepted(BaseModel):
    client_event_id: str
    status: Literal["accepted", "duplicate"]


class PlaybackEventFailed(BaseModel):
    client_event_id: str
    track_id: int
    status: Literal["failed"] = "failed"
    error: str


class PlaybackEventBulkSyncResponse(BaseModel):
    accepted: list[PlaybackEventAccepted]
    failed: list[PlaybackEventFailed]

File: web/src/types/playbackEvent.ts
export type PlaybackEventType =
  | "play"
  | "pause"
  | "resume"
  | "seek"
  | "skip"
  | "complete";

export type PlaybackEventPayload = {
  client_event_id: string;
  track_id: number;
  event_type: PlaybackEventType;
  position_seconds: number;
  duration_seconds: number | null;
  occurred_at: string;
  client: "web";
};

export type PlaybackEventsRequest = {
  events: PlaybackEventPayload[];
};

export type PlaybackEventAccepted = {
  client_event_id: string;
  status: "accepted" | "duplicate";
};

export type PlaybackEventFailed = {
  client_event_id: string;
  track_id: number;
  status: "failed";
  error: string;
};

export type PlaybackEventsResponse = {
  accepted: PlaybackEventAccepted[];
  failed: PlaybackEventFailed[];
};
